sql-data-pipeline/src/main/java/io/cdap/pipeline/sql/app/core/SQLPipelinePlanner.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.pipeline.sql.app.core;
+
+import io.cdap.pipeline.sql.api.template.SQLSink;
+import io.cdap.pipeline.sql.app.core.interfaces.SQLJob;
+
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+
+/**
+ * Creates the {@link SQLJob} which defines the order of queries to be executed.
+ */
+public class SQLPipelinePlanner {
+  private Map<SQLPipelineNode, SQLJob> sinkToJobMap;
+
+  public SQLPipelinePlanner() {
+    sinkToJobMap = new LinkedHashMap<>();
+  }
+
+  public SQLJob plan(SQLPipelineSpec spec) {
+    // Simple validation
+    if (spec == null) {
+      throw new IllegalArgumentException("Input specification cannot be null");
+    }
+
+    // DAG traversal and query ordering logic
+    // Create query sets for each sink by traversing backwards from the sinks
+    // Add dependencies backwards so we don't use temporary tables before creating them
+    // This definition set is to prevent nodes from being defined multiple times
```

Comment:
I don't really understand the algorithm here but it seems backward. Generation of the queries should happen from the start to the end, in order to know the correct table names, properly validate schema, etc.

The Dag class already has a method to get the dag nodes back in topological order, you shouldn't need to write your own traversal logic.

Reply:
The table names and queries can be generated from just knowing the node and previous nodes which connect to it, so those do not require any kind of traversal to perform, but this might change once we add schema verification. The traversal is simply used for ensuring that the final ordering of the queries is in topological order.

I'll look into incorporating the existing DAG class since it already has logic to get the nodes in topological order.